hearth/lib/hearth/connection_pool.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+module Hearth
+  # @api private
+  class ConnectionPool
+    @pools_mutex = Mutex.new
+    @pools = {}
+
+    class << self
+      # @return [ConnectionPool]
+      def for(config = {})
+        @pools_mutex.synchronize do
+          puts "using connection pool for #{config}\n" if @pools[config]
+          puts "creating connection pool for #{config}\n" unless @pools[config]
+          @pools[config] ||= new
+        end
+      end
+
+      # @return [Array<ConnectionPool>] Returns a list of the
+      #   constructed connection pools.
+      def pools
+        @pools_mutex.synchronize do
+          @pools.values
+        end
+      end
+    end
+
+    # @api private
+    def initialize
+      @pool_mutex = Mutex.new
+      @pool = {}
+    end
+
+    # @param [URI::HTTP, URI::HTTPS] endpoint The HTTP(S) endpoint
+    #    to connect to (e.g. 'https://domain.com').
+    # @param [Proc] block A block that returns a new session.
+    # @return [Object, nil]
+    def session_for(endpoint, &block)
```

Comment:
nit - it feels weird for a connection pool to return `session`.  I'd lean towards using `connection` everywhere instead of session.

Reply:
Connection is fine